sylt-tokenizer/src/tokenizer.rs

```diff
@@ -4,42 +4,99 @@ pub use token::Token;
 
 mod token;
 
-pub type PlacedToken = (Token, usize);
-pub type TokenStream = Vec<PlacedToken>;
-
-pub fn string_to_tokens(content: &str) -> TokenStream {
-    let lexer = Token::lexer(&content);
-
-    let mut placed_tokens = lexer.spanned().peekable();
-
-    let mut lined_tokens = Vec::new();
-    let mut line: usize = 1;
-    for (c_idx, c) in content.chars().enumerate() {
-        if let Some((kind, t_range)) = placed_tokens.peek() {
-            if t_range.start == c_idx {
-                let kind = kind.clone();
-                placed_tokens.next();
-                lined_tokens.push((kind, line));
-            }
-        } else {
-            break;
-        }
+#[derive(Debug, Copy, Clone, PartialEq)]
+/// A location in a file containing source code.
+pub struct Span {
+    // TODO(ed): Do this more intelligent, so
+    // we can show ranges. Maybe even go back
+    // to offsets from start of the file.
+    pub line: usize,
+    /// The first column that this Span contains.
+    pub col_start: usize,
+    /// The first column that this Span doesn't contain.
+    pub col_end: usize,
+}
 
-        if c == '\n' {
-            line += 1;
+impl Span {
+    pub fn zero() -> Self {
+        Self {
+            line: 0,
+            col_start: 0,
+            col_end: 0,
         }
     }
+}
 
-    lined_tokens
+#[derive(Debug, PartialEq)]
+pub struct PlacedToken {
+    pub token: Token,
+    pub span: Span,
 }
 
-pub fn file_to_tokens(file: &Path) -> Result<TokenStream, std::io::Error> {
+pub fn string_to_tokens(content: &str) -> Vec<PlacedToken> {
+    // A list containing which char index a specific byte index is at.
+    //
+    // Since &str contains UTF-8, a byte offset (which is what the lexer gives
+    // us) won't necessarily match up with the char index. For example, given
+    // the string "123", '3' has both byte index 3 and char index 3. However, in
+    // the string "ä23", '3' has char index 3 as before, but byte index 4 since
+    // 'ä' contains two bytes.
+    //
+    // This list ensures that the byte offset the lexer gives us can be matched
+    // with a char index. None means that the byte offset points inside a char
+    // which should not be possible.
+    let mut char_at_byte = vec![None; content.len()];
+    for (i, (pos, _)) in content.char_indices().enumerate() {
+        char_at_byte[pos] = Some(i + 1);
+    }
+    // Push a last value since the byte offset end is exclusive.
+    char_at_byte.push(Some(content.chars().count() + 1));
+
+    // We also need to keep track of the current line and which char index the
+    // previous newline was at. Given a char index we can then subtract the last
+    // newline char index and get the column in the current line of the char.
+    let mut line = 1;
+    let mut last_newline = 0;
+
+    Token::lexer(&content)
+        .spanned()
+        .map(|(token, byte_range)| {
+            let is_newline = token == Token::Newline;
```

Comment:
Why not move this check into the if-statement where it's being used?

Reply:
`token` is moved here: https://github.com/FredTheDino/sylt-lang/blob/compiler-spans/sylt-tokenizer/src/tokenizer.rs#L80